cluster-autoscaler/core/utils.go

```diff
@@ -411,10 +411,20 @@ func removeOldUnregisteredNodes(unregisteredNodes []clusterstate.UnregisteredNod
 			}
 			err = nodeGroup.DeleteNodes([]*apiv1.Node{unregisteredNode.Node})
 			if err != nil {
-				klog.Warningf("Failed to remove node %s: %v", unregisteredNode.Node.Name, err)
-				logRecorder.Eventf(apiv1.EventTypeWarning, "DeleteUnregisteredFailed",
-					"Failed to remove node %s: %v", unregisteredNode.Node.Name, err)
-				return removedAny, err
+				_, wasPlaceholder := err.(*cloudprovider.PlaceholderDeleteError)
```

Comment:
I do not believe this is right approach to determine the nature of deleted instance based on error returned from `DeleteNodes()` (especially given the fact that `DeleteNodes` takes multiple nodes as an argument).

Instead we should based the behavior on what we got from `NodeGroup.Nodes()`.
Actually it seems you already have all the needed fields in the interface. The `NodeGroup.Nodes()` returns `InstanceStatus` for each node. And there is `InstanceState` field which can currently be one of `Instance{Running,Creating,Deleting}`. It seems to me that you want to backoff nodegroup if deleted node is in `InstanceCreating` state. If you think you would need to extend the `InstanceStatus` in some way (e.g. introduce more error classes or something - we can discuss it).

Reply:
I don't think the above is right. We want to have feedback about the whole `DeleteNodes()`, we don't need this to be instance-by-instance. The thing is: if you requested 4 new instances, 4 came up successfully, but there's no spot capacity for the 5th, the ASG should be "backed-off" exactly the same as if no instance was started.
Of course, we can set `InstanceState` for each instance, then after the call iterate instances again to check if any is not coming up, but I'd like to introduce a new state for that, like `CreationTimeout`

Reply:
> I don't think the above is right. We want to have feedback about the whole DeleteNodes(), we don't need this to be instance-by-instance. The thing is: if you requested 4 new instances, 4 came up successfully, but there's no spot capacity for the 5th, the ASG should be "backed-off" exactly the same as if no instance was started.

I agree that we want to backoff in such case.

> Of course, we can set InstanceState for each instance, then after the call iterate instances again to check if any is not coming up, but I'd like to introduce a new state for that, like CreationTimeout

`InstanceState` should still be `InstanceCreating`. Maybe the wording is not best but `InstanceCreating` means both:
* instance which is being created and we still do not know what will be outcome of that process
* instance which was created and creation result in a failure

For the latter case the `ErrorInfo` in the `InstanceStatus` is set to non-nil value. And IMO you should use this field to differentiate between successful and failed creation. It seems that you can still use the `OtherErrorClass` when you set `ErrorInfo`.

Reply:
OK, I think I can have a look at that today. BTW, is someone of you going to KubeCon Barcelona next week?

Reply:
THanks. I will be in BCN and bunch of other people from autoscaling team too.

Reply:
There's a problem with this approach. It's not easy to update `Node.InstanceState` within aws provider, as it only keeps track of instances refs (a single string). This is common in the whole provider. Only when `Nodes()` is called, the array of strings (instance ids) is converted to `[]Nodes`, so there's no easy way to keep additional state of nodes without heavily changing interfaces and returned types of aws provider.

Reply:
I understand. Yet I do not believe that changing signature a `DeleteNodes` is the way to go. I do not like the idea of making point changes to general interface, so it is somewhat less of an effort to support it for one specific implementation.

If exposing InstanceState for all the instances in AWS is problematic, maybe you can fairly easily set those to the placeholder nodes only?